apis/r/R/write_seurat.R

```diff
@@ -207,6 +208,215 @@ write_soma.Assay <- function(
   return(ms)
 }
 
+#' @rdname write_soma_seurat_sub
+#'
+#' @section Writing v5 \code{Assays}:
+#' \pkg{Seurat} v5 \code{\link[SeuratObject:Assay5]{Assays}s} are written
+#' out as individual\link[tiledbsoma:SOMAMeasurement]{measurements}:
+#' \itemize{
+#'  \item the layer matrices are written out as
+#'   \link[tiledbsoma:SOMASparseNDArray]{sparse arrays} within the
+#'   \dQuote{\code{X}} group
+#'  \item feature-level metadata is written out as a
+#'   \link[tiledbsoma:SOMADataFrame]{data frame} called \dQuote{\code{var}}
+#' }
+#' Expression matrices are transposed (cells as rows) prior to writing. All
+#' other slots, including results from extended assays (eg. \code{SCTAssay},
+#' \code{ChromatinAssay}) are lost\cr
+#' The following bits of metadata are written in various parts of the measurement
+#' \itemize{
+#'  \item \dQuote{\code{soma_ecosystem_seurat_assay_version}}: written at the
+#'   measurement level; indicates the Seurat assay version.
+#'   Set to \dQuote{\code{v5}}
+#'  \item \dQuote{\code{soma_ecosystem_seurat_v5_default_layers}}: written at
+#'   the \dQuote{\code{X}} group level; indicates the
+#'   \link[SeuratObject:DefaultLayer]{default layers}
+#'  \item \dQuote{\code{soma_ecosystem_seurat_v5_ragged}}: written at the
+#'   \dQuote{\code{X/<layer>}} array level; with a value of
+#'   \dQuote{\code{ragged}}, indicates whether or not the layer is ragged
+#'  \item \dQuote{\code{soma_r_type_hint}}: written at the
+#'   \dQuote{\code{X/<layer>}} array level; indicates the \R class and
+#'   defining package (for S4 classes) of the original layer
+#' }
+#'
+#' @method write_soma Assay5
+#' @export
+#'
+write_soma.Assay5 <- function(
+  x,
+  uri = NULL,
+  soma_parent,
+  ...,
+  ingest_mode = 'write',
+  platform_config = NULL,
+  tiledbsoma_ctx = NULL,
+  relative = TRUE
+) {
+  check_package('SeuratObject', version = '5.0.1')
+  stopifnot(
+    "'uri' must be a single character value" = is.null(uri) ||
+      is_scalar_character(uri),
+    "'soma_parent' must be a SOMACollection" = inherits(
+      x = soma_parent,
+      what = 'SOMACollectionBase'
+    ),
+    "'relative' must be a single logical value" = is_scalar_logical(relative)
+  )
+
+  # Find `shape` if and only if we're called from `write_soma.Seurat()`
+  parents <- unique(sys.parents())
+  idx <- which(vapply_lgl(
+    parents,
+    FUN = function(i) identical(sys.function(i), write_soma.Seurat)
+  ))
+  shape <- if (length(idx) == 1L) {
+    get("shape", envir = sys.frame(parents[idx]))
+  } else {
+    NULL
+  }
```

Comment:
Accessing variables from parent frames seems fragile. Could we just add a `shape` argument?

Reply:
I had thought about that, but I didn't want to expose `shape` in `write_soma()`; this logic is needed for resume-mode #2405, and more specifically for the tests. This functionality isn't used otherwise, which is why I didn't want to expose `shape` for `write_soma()`